### DISCUSSION TITLE
Multi-tenant support

### DISCUSSION THREAD

**Discussion by pjorritsma** [almenscorner/intunecd-monitor#7] [Ideas]
Hello Almenscorner,

Would be awesome if we could Implement multi-tenant support so can manage multi intunecd tenants ;)

**Comment by almenscorner** [comment]
Hi,

I've been interested in adding multi-tenancy support in IntuneCD Monitor for a while :) If you have any ideas or preferences on how this implementation would look, let me know!

**Reply by pjorritsma** [reply]
Helo Almen, will look into it

**Comment by almenscorner** [comment]
@pjorritsma, I have started to think about how an implementation of multi-tenancy could look in IntuneCD Monitor as per attached screenshot. What I'm thinking is that the dashboard view will change when you are using multi tenancy to display the tenants and the update status of them. The pipeline/actions will change to run the IntuneCD tool locally within the container instead and push changes from there to the git repos. I only see a reason to have Update functionality available for the downstream tenants which are using the baseline tenants repo. If backup should be done, it should come from the upstream tenant. What are your thoughts?

![image](https://user-images.githubusercontent.com/78877636/224486958-2db2936e-f398-4a28-ab5a-661c34d0842e.png)

**Reply by Skorfulose** [reply]
I would also love to see multi-tenant support. Our company has three (Development / Integration / Production).

I like the idea of having one blueprint repo, and then staging changes from there to the tenants.
As to the backup, our **usual** flow (for planned changes) of course would be to implement in Development, perform the backup, then update Integration, and finally Production.
But **sometimes**, we have to fix stuff quickly in production without going through the staging environments. In this case, backup from _any_ tenant to the blueprint would be a nice to have.

**Reply by almenscorner** [reply]
Thanks for the insights, I'm mapping this out and testing. Slow going due to time constraints but it's getting there :)

**Comment by almenscorner** [comment]
As of version 2.0 released yesterday, multi-tenant support has been added :) 

#23

**Comment by Cat-Snack-Jack** [comment]
I created my first tenant, tenantabc.onmicrosft.com, and set the URL to github, set the PAT. I marked it as the baseline (as I learned the first one must be a baseline). Seems to run (thought doc creation fails).
I created a new tenant with the same info "tenantabc.onmicrosfot.com" ( i don't recall why I did this), i think it was the name i didn't like, so i made a new (duplicate) tenant.
I then tried to delete the original (baseline), which i was allowed to do, but then i can no longer edit any other tenants that shared the PAT as the secret no longer exists.

The issue as I see it, is any tenants with the same onmicrosoft, share the PAT, thus you can't delete one without messing up both/all of them. And you can't update the PAT as you can no longer edit the "bad" tenant (error below)
I can add a new/3rd tenantasabc.onmicrosoft.com, with the old PAT, but still can't delete the former "same tenants" as the PAT gets deleted regardless. It's an endless loop with multiple entries on the same tenant.

Perhaps the PAT should be truly "per entry" and not "per tenant"

I have to delete the app and re-provision to get around this

Error log/stream
azure.core.exceptions.ResourceNotFoundError: (SecretNotFound) A secret with (name/id) tenantabc-onmicrosoft-com was not found in this key vault. If you recently deleted this secret you may be able to recover it using the correct recovery command.

**Reply by Cat-Snack-Jack** [reply]
I added a 3rd tenant (same tenantabc.onmicrosoft.com) so I could try a few things.
Once I had a valid tenant and PAT (that matched the original tenantabc.onmicrosoft.com), I was able to go into my original tenant entry (the original vaseline tenant) and uncheck the "baseline" so I could choose a different baseline tenant, and now I can't (again) edit any of the three tenants.
